### ISSUE TITLE
In dcgan.py, the parameters of netD are updated as training G?

### ISSUE THREAD

**Issue by thbupt** [apache/incubator-mxnet#12582]
In dcgan.py, the code for updating G is as followings.
print(netD(fake))
        with autograd.record():
            output = netD(fake)
            output = output.reshape((-1, 2))
            errG = loss(output, real_label)
            errG.backward()
        trainerG.step(opt.batch_size)
        print(netD(fake))
I print the output of netD before and after updating G, I found they have different outputs, which means the netD was also updated? In my opinion, only the parameters of G should be updated as excuting trainerG.step(). So what's problem?

**Comment by stu1130** [comment]
Thanks for submitting the issue!
You could also ask the question [link](https://discuss.mxnet.io/) to get more help from community.
@mxnet-label-bot[Question, Example]

**Comment by sandeep-krishnamurthy** [comment]
Hello @thbupt 

I verified that as expected, with trainerG.step(), the Discriminator network (netD) is not updated. Below is the code snippet I used in the training loop under update G network.

```python
        with autograd.record():
            fake = netG(latent_z)
            output = netD(fake).reshape((-1, 1))
            errG = loss(output, real_label)
            errG.backward()
        
        print("Before ")
        print(netD(fake))
        trainerG.step(batch.data[0].shape[0])
        print("After ")
        print(netD(fake))
```

However, please note that you are generating a new fake images here. You are not using fake images generated in Update D network section. So, if you verify netD(fake) before this block, you are actually using different input fake images so the output is different.

Resolving for now. Please reopen if issue still persists or you have any further queries.
